### DISCUSSION TITLE
smartcard authentication

### DISCUSSION THREAD

**Discussion by eugenebliznyuk** [FreeRDP/FreeRDP#7684] [Q&A]
Hi!
Please help! 

It is possible to logon to Windows using certificate on smartcard and suppling pin-code with option "+password-is-pin" and parameter "/p:<pin>". 

But is it possible to repeat this while there are two or more certificates witn private keys on smartcard? How can I point to one or another certificate starting xfreerdp from cli?

**Comment by hardening** [comment]
With which kind of security (TLS or NLA ) ?

**Reply by eugenebliznyuk** [reply]
I tried TLS and RDP. As I know, NLA security doesn't work with smartcard authentication (cause it is needed to provide login and password).

**Reply by akallabeth** [reply]
With `TLS/RDP` security you are presented with the default windows login screen and can choose the device/certificate there.
`FreeRDP` is just a wrapper forwarding smartcard calls in that case.
@hardening did some nice work on getting `NLA` support for smartcard auth upstream, but that is currently `bleeding edge` and you might encounter some hickups/obstacles along the way.

**Reply by Demonist** [reply]
Why is it possible to skip logon screen using `TLS/RDP` with one certificate on smartcard but with several certificates I have to manually choose user on logon screen and enter pin-code again? 
If I set `/u:` option it doesn't work. I still need to choose a user.

Is it a restriction of FreeRDP or windows logon screen?

**Reply by hardening** [reply]
That's definitely a restriction of the TLS mode. With a NLA logon, you establish the connection (Kerberos PKINIT) with the certificate you've chosen, and there's only a verification once the transport is established from the server.
With TLS, it's basically the greeter's GUI that will do the job and will let you select the certificate. With TLS AFAICT there's no way to specify the certificate to use.

**Comment by rahra** [comment]
> Hi! Please help!
> 
> It is possible to logon to Windows using certificate on smartcard and suppling pin-code with option "+password-is-pin" and parameter "/p:".
> 
> But is it possible to repeat this while there are two or more certificates witn private keys on smartcard? How can I point to one or another certificate starting xfreerdp from cli?

Yes it is, at least with xfreerdp3.
NLA works, Smartcard-Login works, a combination of both works, and having a smartcard with multiple private keys on it works.
I did it with a Yubikey 5. 

Make sure that Kerberos works as explained in this article: https://scriptech.io/linux-enable-smartcard-authentication-against-active-directory-and-generate-tgt-using-pkinit/

Everything else is then straight forward: `xfreerdp3 /v:your.terminal.server /smartcard-logon /sec:nla`

**Comment by hardening** [comment]
There's also https://www.hardening-consulting.com/en/posts/20231004-smartcard-logon.html which addresses the smartcard logon thing with a more limited range: just doing smartcard auth with RDP not your full linux session.
